src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -16,230 +16,293 @@ import io.iohk.ethereum.network.discovery.PeerDiscoveryManager
-                                                    managerState.outgoingPeers.size < peerConfiguration.maxOutgoingPeers)
-    } yield connectionNumberValidation
+    val connectionHandled = isConnectionHandled(remoteAddress, pendingPeers, peers)
+    val isOutgoingPeersNotMaxValue = countOutgoingPeers(peers) < peerConfiguration.maxOutgoingPeers
 
-    validatedConnection match {
-      case Right(addr) =>
-        val peer = createPeer(addr, incomingConnection = false)
+    val validConnection = for {
+      validHandler <- validateConnection(remoteAddress, OutgoingConnectionAlreadyHandled(uri), connectionHandled)
+      validNumber <- validateConnection(validHandler, MaxOutgoingConnections, isOutgoingPeersNotMaxValue)
+    } yield validNumber
+
+    validConnection match {
+      case Right(address) =>
+        val (peer, newPendingPeers, newPeers) = createPeer(address, incomingConnection = false, pendingPeers, peers)
         peer.ref ! PeerActor.ConnectTo(uri)
+        context become listen(newPendingPeers, newPeers)
 
       case Left(error) => handleConnectionErrors(error)
     }
   }
 
-  def handleCommonMessages: Receive = {
+  private def isConnectionHandled(remoteAddress: InetSocketAddress, pendingPeers: PeerMap, peers: PeerMap): Boolean =
+    !(peers ++ pendingPeers).values.map(_.remoteAddress).toSet.contains(remoteAddress)
+
+  private def outgoingPeersAddresses(peers: PeerMap): Set[InetSocketAddress] =
+    peers.filter{ case (_, p) => !p.incomingConnection }.values.map(_.remoteAddress).toSet
+
+  private def countOutgoingPeers(peers: PeerMap): Int = peers.count{ case (_, p) => !p.incomingConnection }
+
+  private def countIncomingPeers(peers: PeerMap): Int = peers.count{ case (_, p) => p.incomingConnection }
+
+  def handleCommonMessages(pendingPeers: PeerMap, peers: PeerMap): Receive = {
     case GetPeers =>
-      getPeers().pipeTo(sender())
+      getPeers(peers.values.toSet).pipeTo(sender())
 
-    case SendMessage(message, peerId) if managerState.peers.contains(peerId) =>
-      val peer = managerState.peers(peerId)
-      peer.ref ! PeerActor.SendMessage(message)
+    case SendMessage(message, peerId) if peers.contains(peerId) =>
+      peers(peerId).ref ! PeerActor.SendMessage(message)
 
     case Terminated(ref) =>
-      val maybePeerId = managerState.findTerminatedPeer(ref)
-      maybePeerId.foreach { peerId =>
+      val terminatedPeers = terminatedPeersIds(ref, pendingPeers, peers)
+
+      terminatedPeers.foreach{ peerId =>
         peerEventBus ! Publish(PeerDisconnected(peerId))
-        managerState.removeTerminatedPeer(peerId)
       }
+      context unwatch ref
+      context become listen(pendingPeers -- terminatedPeers, peers -- terminatedPeers)
 
-    case IncomingConnectionHandshakeSuccess(peerId, peer) =>
-      if (managerState.incomingPeers.size >= peerConfiguration.maxIncomingPeers) {
+    case IncomingConnectionHandshakeSuccess(peer) =>
```

Comment:
I think we can get rid of that case if we are already doing refactoring e.g;
1. We can subscribe to `peerEventBus ! Subscribe(SubscriptionClassifier.PeerHandshaked)`  during action creation
2. And then cover case `case PeerEvent.PeerHandshakeSuccessful(peer, _) if peer.incomingConnection` 
and then we cen get rid of 
3. https://github.com/input-output-hk/mantis/pull/547/files#diff-d88dc0e9076bb9b1c58875187ebcd4acR230

I think that solution will be clearer, and use our existing pub-sub infrastrucure. Wdyt ?

Reply:
check now if you had something like this in mind